### DISCUSSION TITLE
Why does the nv_open_q process in my system take up too much CPU without running additional processes outside the system?

### DISCUSSION THREAD

**Discussion by Turbolento** [NVIDIA/open-gpu-kernel-modules#615] [Q&A]
### **Open Source Driver:**  v550.40.07

**I will provide more information if necessary.**

![image](https://github.com/NVIDIA/open-gpu-kernel-modules/assets/30725265/0bce00af-1095-458a-b30b-487b4d984eb8)
![image](https://github.com/NVIDIA/open-gpu-kernel-modules/assets/30725265/3ffc1293-c2bc-4357-9c19-ba0ad51fa7ba)

**Comment by Turbolento** [comment]
Before installing the Open Source Driver, the CPU usage was below 5%. After installing the Driver, the usage has been around 25%.

**Reply by aritger** [reply]
That is surprising.  open_q was added fairly recently to better parallelize user-space processes opening /dev/nvidia* files, but it should be idle the vast majority of the time.

What else is the system doing when you see this?  E.g., do you have an X server running?  Or wayland compositor?  Or nvidia-persistenced?

An nvidia-bug-report.log might help, too.

**Comment by Turbolento** [comment]
@aritger Thank you for your answer. After 24 hours, the CPU usage should have returned to the normal range. However, I will continue to observe in the future. If the CPU abnormality occurs again, I will record and upload nvidia-bug-report.log

![image](https://github.com/NVIDIA/open-gpu-kernel-modules/assets/30725265/7d72dbc6-bf54-4d22-85b2-c02b0a575081)

**Comment by x3ek** [comment]
I'm on UNRAID 6.12.10 and I am noticing this going on right now. It's pegging 3 of my logical processors.
![image](https://github.com/NVIDIA/open-gpu-kernel-modules/assets/6032840/9fd92aab-7edd-48cd-8253-b581c352d89f)

**Reply by x3ek** [reply]
![image](https://github.com/NVIDIA/open-gpu-kernel-modules/assets/6032840/fa84b895-7970-4a68-b893-907ee4eef322)

There shouldn't be anything using this GPU right now; I do not see any apps utilizing it in the GPU stats right now, no one is watching Plex videos where I have it configured to be used for encoding right now.

**Reply by x3ek** [reply]
[nvidia-bug-report.log.gz](https://github.com/NVIDIA/open-gpu-kernel-modules/files/15136110/nvidia-bug-report.log.gz)
The nvidia-bug-report.log for my system. I actually couldn't kill that process for some reason so I'm about to reboot the system.

**Comment by Masses3581** [comment]
I could copy&paste everything from @xeekworx. I have the same version of Unraid Server 6.12.10, same NVIDIA driver version 550.67, same problems.... :-( The nv_open_q thread is boosting up my CPU system, normally it should stay quiet.
The only difference i see  ist: i am using emby server instead of plex.
[nvidia-bug-report.log.gz](https://github.com/NVIDIA/open-gpu-kernel-modules/files/15159526/nvidia-bug-report.log.gz)

I hope this helps maybe to identify the problem source. Thank you for help.

**Comment by mtijanic** [comment]
As Andy says, this is really surprising, probably a case of some userspace process misbehaving. We'll dig through the attached bug reports, thanks!

In the meantime, could someone please try using the [`NVreg_EnableNonblockingOpen=0`](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/550.78/kernel-open/nvidia/nv-reg.h#L825-L840) module param and report back whether it makes the issue go away?

**Reply by JsGraaf** [reply]
Fixed it for me. For reference, I'm on Fedora 40 with 6.8.10-300.fc40.x86_64 and Nvidia kernel 550.78 with a GTX1650 Max-Q

**Comment by amrit1711** [comment]
Hi All,
Request you to please trace open calls with bpftrace as per below example.
You can download the latest binary from https://github.com/bpftrace/bpftrace/releases.

bpftrace -e 'kprobe:nvidia_open { printf("[%llu] nvidia_open by %s\n", nsecs, comm); }'

**Comment by kbareis** [comment]
For folks on Unraid, I think I figured it out. My `nv_open_q` only pegs 50% CPU when I am on the dashboard. This is likely due to having the GPU widget enabled so I can see process, memory usage, and such. What is funny is as soon as I move to another tab, the process drops back down to almost 0.

**Reply by luyufan498** [reply]
Same to me,  seems that the dashboard is the key ...

**Comment by luyufan498** [comment]
![image](https://github.com/user-attachments/assets/887340c8-dc7d-44ed-be56-b9c8655bb40a)

The nv_open_q  uses  50% of  8700K..

switch to the other panel does fix this....

**Comment by CorneliusCornbread** [comment]
Seeing this happen with Unraid as well, nv_open_q using tons of CPU usage while Plex is idle. Insanely obnoxious, driver version 560.28.03, Unraid 6.12.11

**Comment by ich777** [comment]
The maintainer from the Nvidia Driver plugin for Unraid here... :)

Can you guys please check the interval in your GPU Statistics plugin and see to what it's set?
If you only see this behaviour on the Unraid Dashboard then it's most certainly caused by the GPU Statistics plugin since if you set the interval in GPU Statistics to 2 seconds it will call `nvidia-smi` every 2 seconds which causes spikes (this is a known thing), I would recommend to set the interval to 5 seconds or even higher.

Please also note that any open Unraid Dashboard page, regardless on which Device, will cause these spikes, so don't forget to check your other Devices if maybe a Unraid Dashboard page is open somewhere.

**Reply by mtijanic** [reply]
Hey there! `nvidia-smi` is indeed one of the rare NV-authored things that passes `O_NONBLOCK` when opening the device file, which triggers the `nv_open_q` to do something. So running `nvidia-smi` in a loop _could_ cause this, but 2 seconds is a long enough interval that it wouldn't register at all. Are you sure it is actually 2 seconds and not something like 2 milliseconds?

Can you also check if you do `open("/dev/nvidia*", O_NONBLOCK)` anywhere else?

**Reply by ich777** [reply]
The GPU Statistics plugin only calls `nvidia-smi` every 2 seconds to get the stats as xml.

**Reply by mtijanic** [reply]
What about all the other calls to `nvidia-smi`? From a quick skim through your code I see it's used to query the driver version and to list out the GPUs in the system and possibly other things. Can you double-check how often the process is spawned? Maybe replace all calls to it with a custom script (or put a script named `nvidia-smi` in the $PATH at higher priority).

The only explanation I can think of currently is that something in the dashboard is running `nvidia-smi` (or some other `libnvidia-ml.so` wrapper) in a _very_ tight loop.

**Reply by ich777** [reply]
Sorry but this is not my code.
GPU Statistics is another plugin and not the driver itself.

I was just here to say that this is most certainly caused by the GPU Statistics plugin.

My code is just called when a user visits the plugin page to display mi imal information about the card(s).

**Reply by mtijanic** [reply]
My apologies. Yeah, I now see there's a second repo that also calls `nvidia-smi` in a few places. It's really hard for me to figure out exactly how often it's called from the code alone, but I really expect the issue is that it just runs way too often.

It certainly doesn't seem like a driver bug, so unless more evidence shows up to the contrary, don't expect a driver update to fix it.

We could theoretically make the behavior of [`NVreg_EnableNonblockingOpen=0`](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/550.78/kernel-open/nvidia/nv-reg.h#L825-L840) be the default if there is only one GPU in the system (since the benefit of it is only on multi-GPU), but I'd much rather see the root cause fixed as it likely has other perf impacts that aren't as easy to spot as 100% CPU usage.

**Reply by BlueGoliath** [reply]
You shouldn't be invoking `nvidia-smi`. It is not and never was intended to be used in the way you're using it for.

From [Nvidia's own website](https://docs.nvidia.com/deploy/nvidia-smi/index.html):

>The output of NVSMI is not guaranteed to be backwards compatible. However, both NVML and the Python bindings are backwards compatible, and should be the first choice when writing any tools that must be maintained across NVIDIA driver releases.

I don't know what is up with web UI monitoring apps / plugins using garbage polling methods like bash scripts or Python command line invoking and parsing but it needs to stop.

I'm the developer of a Java Nvidia GPU utility app that reports an extreme amounts of data(120+ base, over 200+ with plugins), 100 histogram charts basically all the time, and 100ms update intervals and I have never had this issue.

**Reply by ich777** [reply]
@mtijanic no worries at all.

I think this question is now solved, the root cause is found and it‘s definitely not caused by the driver itself. Definitely nothing that needs to be changed from your side in my oppinion.
I just wanted to chime in because I didn‘t want that convo be longer than necessary.

@BlueGoliath thanks for your input but I think you‘ve should have read the last parts oft this convo since I can‘t do anything about it because I just maintain the Nvidia driver for Unraid and not the GPU Statistics plugin.
Maybe you could point me to your app/repo that I can give feedback to the developer from the other plugin so that he can take a look at it.

**Reply by BlueGoliath** [reply]
> @BlueGoliath thanks for your input but I think you‘ve should have read the last parts oft this convo since I can‘t do anything about it because I just maintain the Nvidia driver for Unraid and not the GPU Statistics plugin.

You dropped into an existing conversation and make it sound like you're someone involved in the project showcasing the issue when you're not. You also bring up drivers and plugins as-if anyone is supposed to know what those terms mean in the context of Unraid.

But whatever, i'm assuming you're referring to:

https://forums.unraid.net/topic/89453-plugin-gpu-statistics/

Which links to this Github:

https://github.com/b3rs3rk/gpustat-unraid

It uses PHP to get GPU info via `nvidia-smi`:

https://github.com/b3rs3rk/gpustat-unraid/blob/00851787ff159cf811f916f0ae5c9bab80c61670/src/gpustat/usr/local/emhttp/plugins/gpustat/lib/Nvidia.php#L37

This is not how anyone should be doing this. File a bug on their Github.

**Reply by ich777** [reply]
@BlueGoliath thanks for your input again and my appoligies.

Sorry but that‘s the wrong repo that you‘ve linked, I‘ll report that and link this convo to him.

I don‘t know why most poeple come in that offensive to such conversations, however I think this is solved for now and I will forward this convo to the developer from GPU Statistsics.

Oh I see you still haven‘t linked your repo so that he can look into how you are doing it.

**Reply by BlueGoliath** [reply]
>Sorry but that‘s the wrong repo that you‘ve linked

How? It's on the first post and under the same username.

>Oh I see you still haven‘t linked your repo so that he can look into how you are doing it.

This is partially why. Nvidia has made it crystal clear for **MANY** years now that  you're only supposed to be using `nvidia-smi` for quick command line reading/control and scripting and not real-world applications and yet despite that people continue to do it and are surprised when things break or the performance is poor.

Nvidia literally says:

>Note that much of the functionality of NVSMI is provided by the underlying NVML C-based library. See the NVIDIA developer website link below for more information about NVML. NVML-based python bindings are also available.

and

>The output of NVSMI is not guaranteed to be backwards compatible. However, both NVML and the Python bindings are backwards compatible, and should be the first choice when writing any tools that must be maintained across NVIDIA driver releases.

I don't know what needs explaining here.

**Reply by ich777** [reply]
Thanks @BlueGoliath

**Reply by SimonFair** [reply]
> Which links to this Github:
> 
> https://github.com/b3rs3rk/gpustat-unraid

I took over the plugin, but the original support post still has the original developer. Here is the current repo https://github.com/SimonFair/gpustat-unraid

The process still run nvsmi. Can you provide guidance on correct process to get the stats and I can look to see if the plugin can support that way.

**Reply by mtijanic** [reply]
What @BlueGoliath says about NVML is true, however we do recognize that running `nvidia-smi` and parsing the result is often much easier. We are looking into ways of making that flow easier to use and less prone to unexpected perf issues.

@SimonFair, thanks! Looking at the linked code the usage of `nvidia-smi` seems reasonable, but I think whatever is calling into that PHP might be doing so way more than you'd expect. There seem to be two different .page files that use different calls to `setInterval()` to set up the refresh rate, and I got lost trying to follow it all in an unfamiliar codebase. I'd suggest you just double-check how often you are _actually_ calling nvidia-smi (both for `getStatistics()` and for `getInventory()` families of functions). There might be something in the dashboard infrastructure itself that is calling you more than you'd expect.

My suggestion is to make a wrapper script `nvidia-smi-cnt.sh` that looks something like:
```sh
#!/bin/bash
if [[ ! -f "/tmp/nvsmi-calls-cnt" ]] || ! [[ $(cat "/tmp/nvsmi-calls-cnt") =~ ^[0-9]+$ ]]; then
    echo "0" > "/tmp/nvsmi-calls-cnt"
fi
CNT=$(echo "$(cat "/tmp/nvsmi-calls-cnt") + 1" | bc)
echo "$CNT" > "/tmp/nvsmi-calls-cnt"
echo "Total nvidia-smi calls: $CNT"

nvidia-smi $@
```
and then switch to calling that instead, then see how fast the counter gets incremented.

By the way, @ich777 , looking at [your repo uses of nvidia-smi](https://github.com/search?q=repo%3Aich777%2Funraid-nvidia-driver%20nvidia-smi&type=code), three out of the four can just be replaced with `cat /sys/module/nvidia/version`. For the last one, using nvidia-smi is reasonable if you need the UUID, I don't think it's easily accessible via sysfs.

**Reply by SimonFair** [reply]
> getInventory()

This only gets call from the settings page file. the dash board will call the command every x seconds depending on the refresh the user specifies. My test system has multiple gpus

**Reply by SimonFair** [reply]
> You can find the header [here](https://github.com/NVIDIA/nvidia-settings/blob/main/src/nvml.h). While I hate Python, it would probably be easiest since bindings already exist unless you're willing to put work into a C server or some other custom solution.

thanks will review I dont think python is the best approach but will look into the api. Thanks for the pointers.

**Comment by PhyoeBlitz** [comment]
I'll be too late when i write this.By the way,
did you open nvidia-smi like i did?
when i run nvidia-smi with
`watch -n 1 nvidia-smi`
it consumes 50% of cpu. When I close it, the CPU usage drops to 0%.
